File: 0139-word-break/0139-word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        if s is None:
            return False
        
        dp = [False for i in range(len(s) + 1)]
        dp[0] = True

        hashSet = set(wordDict)

        for i in range(len(s) + 1):
            j = 0
            while j < i:
                if dp[j] and s[j : i] in hashSet:
                    dp[i] = True
                    break
                j+= 1
        
        return dp[-1]

#DP
# Time Complexity: O(N^2)
#Space Complexity: O(n)

File: 0139-word-break/0139-word-break.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        if(s == null || s.length() == 0 || wordDict == null || wordDict.size() == 0){
            return false;
        }
        HashSet<String> set = new HashSet<>(wordDict);
        boolean[] dp = new boolean[s.length() + 1];
        dp[0] = true;

        for(int i = 0; i< dp.length; i++){
            for(int j = 0; j < i; j++){
                if(dp[j] && set.contains(s.substring(j, i))){
                    dp[i] = true;
                    break;
                }
            }
        }

        return dp[s.length()];
    }
}

// #DP
// # Time Complexity: O(n^2)
// #Space Complexity: O(n)
